### ISSUE TITLE
Coinbase Transaction creation failed at 144 Block Height

### ISSUE THREAD

**Issue by TrustHenry** [bosagora/agora#2461]
```

agora_1     | 2021-09-07 06:43:45,463 Info [agora.consensus.Ledger] - Completed externalization of block #142
agora_1     | 2021-09-07 06:43:45,469 Info [agora.consensus.protocol.Nominator] - agora.consensus.protocol.Nominator.Nominator.nominate(): Tx set triggered new nomination
agora_1     | 2021-09-07 06:43:50,447 Info [agora.consensus.protocol.Nominator] - Nominating { tx_set: [], enrolls: [], missing_validators: [], time_offset: 21537830 } at 1630997030
agora_1     | 2021-09-07 06:43:50,447 Info [agora.consensus.protocol.Nominator] - agora.consensus.protocol.Nominator.Nominator.nominate(): Proposing tx set for slot 143
agora_1     | 2021-09-07 06:43:50,451 Info [agora.consensus.protocol.Nominator] - Externalized consensus data set at 143: { tx_set: [], enrolls: [], missing_validators: [], time_offset: 21537830 }
agora_1     | 2021-09-07 06:43:50,453 Info [agora.consensus.Ledger] - Beginning externalization of block #143
agora_1     | 2021-09-07 06:43:50,453 Info [agora.consensus.Ledger] - Transactions: 0 - Enrollments: 0
agora_1     | 2021-09-07 06:43:50,453 Info [agora.consensus.Ledger] - Validators: Active: 1 - Signing: 1 - Slashed: []
stoa_1      | [Stoa] 2021-09-07T06:43:50.481Z info Saved a block with block height of 143
stoa_1      | [Stoa] 2021-09-07T06:43:50.482Z info Emitted new Block: [object Object]
stoa_1      | [Stoa] 2021-09-07T06:43:50.482Z info Emitted new Transactions:
stoa_1      | [Stoa] 2021-09-07T06:43:50.483Z info Emitted Updated BOA stats:  [object Object]
toolsdev_1  | AutoSend: Status = false; Hash = undefined
agora_1     | 2021-09-core.exception.AssertError@source/agora/utils/Log.d(105): Orphan format specifier: %s
stoa_1      | [Stoa] 2021-09-07T06:43:55.455Z info Saved a pre-image utxo : 0x210f6551d648a4da, hash : 0xe6ee459c8f003a0a, pre-image height : 149
agora_1     | ----------------
agora_1     | source/agora/node/main.d:276 nothrow void agora.node.main.handleAssertion(immutable(char)[], ulong, immutable(char)[]) [0x56214612c790]
agora_1     | source/agora/utils/Log.d:87 nothrow @safe void agora.utils.Log.Logger.format!(agora.crypto.Key.PublicKey).format(ocean.util.log.ILogger.ILogger.Level, const(char)[], agora.crypto.Key.PublicKey) [0x562146101b60]
agora_1     | source/agora/utils/Log.d:75 nothrow @safe void agora.utils.Log.Logger.error!(agora.crypto.Key.PublicKey).error(const(char)[], agora.crypto.Key.PublicKey) [0x562146101b20]
agora_1     | source/agora/consensus/Ledger.d:613 nothrow @safe agora.consensus.data.Transaction.Transaction agora.consensus.Ledger.Ledger.getCoinbaseTX(in agora.common.Types.Height) [0x5621461c7110]
agora_1     | source/agora/consensus/Ledger.d:1391 @safe void agora.consensus.Ledger.ValidatingLedger.prepareNominatingSet(out agora.consensus.protocol.Data.ConsensusData, ulong, ulong) [0x5621461aa0a0]
agora_1     | source/agora/consensus/protocol/Nominator.d:361 @safe bool agora.consensus.protocol.Nominator.Nominator.prepareNominatingSet(out agora.consensus.protocol.Data.ConsensusData) [0x5621461a9f10]
agora_1     | source/agora/consensus/protocol/Nominator.d:427 @safe void agora.consensus.protocol.Nominator.Nominator.checkNominate() [0x5621461a9c70]
agora_1     | submodules/vibe-core/source/vibe/core/core.d:1212 nothrow @safe void vibe.core.core.setTimer(core.time.Duration, void delegate(), bool).__lambda4() [0x56214691f480]
agora_1     | submodules/vibe-core/source/vibe/core/core.d:1270 nothrow @safe void vibe.core.core.createTimer(void delegate() nothrow @safe).C.opCall(vibe.core.core.Timer).__lambda2(vibe.core.core.Timer) [0x56214691f600]
agora_1     | submodules/vibe-core/source/vibe/core/task.d:708 nothrow void vibe.core.task.TaskFuncInfo.set!(void delegate(vibe.core.core.Timer) nothrow @safe, vibe.core.core.Timer).set(ref void delegate(vibe.core.core.Timer) nothrow @safe, ref vibe.core.core.Timer).callDelegate(ref vibe.core.task.TaskFuncInfo) [0x562146961f70]
agora_1     | submodules/vibe-core/source/vibe/core/task.d:737 void vibe.core.task.TaskFuncInfo.call() [0x56214695f8b0]
agora_1     | submodules/vibe-core/source/vibe/core/task.d:403 nothrow void vibe.core.task.TaskFiber.run() [0x56214695edb0]
agora_1     | ??:? fiber_entryPoint [0x7fa0b231d0a0]
```

It was discovered during testing with a single agora node.

It used this agora docker image.
https://hub.docker.com/layers/bosagora/agora/latest/images/sha256-fb24b4043fbb7b364eac0efa8b7a08a85c6e730d8f0fc010e91f27c8f343ebb3?context=explore

config
```
consensus:
  block_interval:
    seconds: 5
  validator_cycle: 20
````

**Comment by Geod24** [comment]
Confirmed. The Exception we're seeing will be fixed in https://github.com/bosagora/agora/pull/2463
The exception might be hiding the real error, or the `log.error` call might be wrong.

**Comment by TrustHenry** [comment]
Error after applied #2463

```
agora_1     | 2021-09-07 09:02:12,366 Info [agora.consensus.Ledger] - Completed externalization of block #143
agora_1     | 2021-09-07 09:02:12,373 Info [agora.consensus.protocol.Nominator] - agora.consensus.protocol.Nominator.Nominator.nominate(): Tx set triggered new nomination
agora_1     | 2021-09-07 09:02:17,350 Info [agora.network.Manager] - Net time offsets: [{ key: boa1xrval7gwhjz4k9raqukcnv2n4rl4fxt74m2y9eay6l5mqdf4gntnzhhscrh, node_time: 1631005337, req_delay: 0, offset: 0 }]
agora_1     | 2021-09-07 09:02:17,351 Error [agora.consensus.Ledger] - Zero valued Coinbase output for key boa1xqcmmns5swnm03zay5wjplgupe65uw4w0dafzsdsqtwq6gv3h3lcz24a8ch
agora_1     | 2021-09-07 09:02:17,351 Error [agora.consensus.Ledger] - Zero valued Coinbase output for key boa1xrval7gwhjz4k9raqukcnv2n4rl4fxt74m2y9eay6l5mqdf4gntnzhhscrh
agora_1     | 2021-09-07 09:02:17,351 Error [agora.consensus.Ledger] - Zero valued Coinbase output for key boa1xqcmmns5swnm03zay5wjplgupe65uw4w0dafzsdsqtwq6gv3h3lcz24a8ch
agora_1     | 2021-09-07 09:02:17,351 Error [agora.consensus.Ledgercore.exception.AssertError@source/agora/consensus/Ledger.d(1175)
stoa_1      | [Stoa] 2021-09-07T09:02:17.360Z info Saved a pre-image utxo : 0x210f6551d648a4da, hash : 0xe6ee459c8f003a0a, pre-image height : 149
agora_1     | ----------------
agora_1     | source/agora/node/main.d:276 nothrow void agora.node.main.handleAssertion(immutable(char)[], ulong, immutable(char)[]) [0x55d9dc7f8590]
agora_1     | source/agora/consensus/Ledger.d:1140 nothrow @safe immutable(char)[] agora.consensus.Ledger.Ledger.getValidTXSet(in ref agora.consensus.protocol.Data.ConsensusData, ref agora.consensus.data.Transaction.Transaction[]) [0x55d9dc87b0d0]
agora_1     | source/agora/consensus/Ledger.d:659 nothrow @trusted immutable(char)[] agora.consensus.Ledger.Ledger.validateConsensusData(in ref agora.consensus.protocol.Data.ConsensusData, in uint[]) [0x55d9dc87a3c0]
agora_1     | source/agora/consensus/protocol/Nominator.d:361 @safe bool agora.consensus.protocol.Nominator.Nominator.prepareNominatingSet(out agora.consensus.protocol.Data.ConsensusData) [0x55d9dc879ea0]
agora_1     | source/agora/consensus/protocol/Nominator.d:427 @safe void agora.consensus.protocol.Nominator.Nominator.checkNominate() [0x55d9dc879c00]
agora_1     | submodules/vibe-core/source/vibe/core/core.d:1212 nothrow @safe void vibe.core.core.setTimer(core.time.Duration, void delegate(), bool).__lambda4() [0x55d9dcffdff0]
agora_1     | submodules/vibe-core/source/vibe/core/core.d:1270 nothrow @safe void vibe.core.core.createTimer(void delegate() nothrow @safe).C.opCall(vibe.core.core.Timer).__lambda2(vibe.core.core.Timer) [0x55d9dcffe170]
agora_1     | submodules/vibe-core/source/vibe/core/task.d:708 nothrow void vibe.core.task.TaskFuncInfo.set!(void delegate(vibe.core.core.Timer) nothrow @safe, vibe.core.core.Timer).set(ref void delegate(vibe.core.core.Timer) nothrow @safe, ref vibe.core.core.Timer).callDelegate(ref vibe.core.task.TaskFuncInfo) [0x55d9dd040ae0]
agora_1     | submodules/vibe-core/source/vibe/core/task.d:737 void vibe.core.task.TaskFuncInfo.call() [0x55d9dd03e420]
agora_1     | submodules/vibe-core/source/vibe/core/task.d:403 nothrow void vibe.core.task.TaskFiber.run() [0x55d9dd03d920]
agora_1     | ??:? fiber_entryPoint [0x7fbd816b50a0]
```

**Comment by hewison-chris** [comment]
I think this can only happen if there are no transactions in any blocks from height 1 -> 144.  It is a bug still but can you confirm if this is the case in this test run.

**Comment by TrustHenry** [comment]
> I think this can only happen if there are no transactions in any blocks from height 1 -> 144. It is a bug still but can you confirm if this is the case in this test run.

Yes, that's right.
No transaction occurred after the test was executed.
It was before the transaction occurred.

**Comment by hewison-chris** [comment]
Thanks for checking. This problem would not occur when Block rewards are added to the payouts as then there will always be payouts to the enrolled validators who signed the blocks. I also added a check so that an output with zero Amount will not be added to the `Coinbase` transaction. I will raise a PR later today for this.

**Comment by hewison-chris** [comment]
I can confirm this is fixed once we merge https://github.com/bosagora/agora/pull/2472. 
Locally I ran a single agora node with `payout_period` of 5 and did not send any transactions.
At block #10, and every 5 blocks after, I see a single `Coinbase` `transaction`.

**Comment by TrustHenry** [comment]
@hewison-chris Does it look related?


agora v0.21 over the commit cb3125a65d35202d3a04912112854e9a4dae0732
In this version, an error occurred while creating 432 height block in the single node test.
```
2021-09-17 02:40:54,124 Info [agora.consensus.protocol.Nominator] - agora.consensus.protocol.Nominator.Nominator.nominate(): Proposing tx set for slot 431
2021-09-17 02:40:54,128 Info [agora.consensus.protocol.Nominator] - Externalized consensus data set at 431: { tx_set: [], enrolls: [], missing_validators: [], time_offset: 22387254 }
2021-09-17 02:40:54,129 Info [agora.consensus.Ledger] - Beginning externalization of block #431
2021-09-17 02:40:54,129 Info [agora.consensus.Ledger] - Transactions: 0 - Enrollments: 0
2021-09-17 02:40:54,129 Info [agora.consensus.Ledger] - Validators: Active: 1 - Signing: 1 - Slashed: []
2021-09-17 02:40:54,134 Info [agora.consensus.Ledger] - Completed externalization of block #431
2021-09-17 02:40:54,139 Info [agora.consensus.protocol.Nominator] - agora.consensus.protocol.Nominator.Nominator.nominate(): Tx set triggered new nomination
2021-09-17 02:41:49,121 Info [agora.network.Manager] - Net time offsets: [{ key: boa1xrval7gwhjz4k9raqukcnv2n4rl4fxt74m2y9eay6l5mqdf4gntnzhhscrh, node_time: 1631846509, req_delay: 0, offset: 0 }]
2021-09-17 02:42:49,122 Info [agora.network.Manager] - Net time offsets: [{ key: boa1xrval7gwhjz4k9raqukcnv2n4rl4fxt74m2y9eay6l5mqdf4gntnzhhscrh, node_time: 1631846569, req_delay: 0, offset: 0 }]
2021-09-17 02:43:49,121 Info [agora.network.Manager] - Net time offsets: [{ key: boa1xrval7gwhjz4k9raqukcnv2n4rl4fxt74m2y9eay6l5mqdf4gntnzhhscrh, node_time: 1631846629, req_delay: 0, offset: 0 }]
2021-09-17 02:43:54,136 Error [agora.consensus.Ledger] - Zero valued Coinbase output for key boa1xqcmmns5swnm03zay5wjplgupe65uw4w0dafzsdsqtwq6gv3h3lcz24a8ch
payouts=[ boa1xqcmmns5swnm03zay5wjplgupe65uw4w0dafzsdsqtwq6gv3h3lcz24a8ch: 0, boa1xrval7gwhjz4k9raqukcnv2n4rl4fxt74m2y9eay6l5mqdf4gntnzhhscrh: 0 ]
2021-09-17 02:43:54,136 Error [agora.consensus.Ledger] - Zero valued Coinbase output for key boa1xrval7gwhjz4k9raqukcnv2n4rl4fxt74m2y9eay6l5mqdf4gntnzhhscrh
payouts=[ boa1xqcmmns5swnm03zay5wjplgupe65uw4w0dafzsdsqtwq6gv3h3lcz24a8ch: 0, boa1xrval7gwhjz4k9raqukcnv2n4rl4fxt74m2y9eay6l5mqdf4gntnzhhscrh: 0 ]
core.exception.AssertError@source/agora/consensus/Ledger.d(677): payouts=[boa1xqcmmns5swnm03zay5wjplgupe65uw4w0dafzsdsqtwq6gv3h3lcz24a8ch:0, boa1xrval7gwhjz4k9raqukcnv2n4rl4fxt74m2y9eay6l5mqdf4gntnzhhscrh:0]
----------------
source/agora/node/main.d:276 nothrow void agora.node.main.handleAssertion(immutable(char)[], ulong, immutable(char)[]) [0x55702785fbd0]
source/agora/consensus/Ledger.d:622 nothrow @safe agora.consensus.data.Transaction.Transaction agora.consensus.Ledger.Ledger.getCoinbaseTX(in agora.common.Types.Height) [0x557027613400]
source/agora/consensus/Ledger.d:1425 @safe void agora.consensus.Ledger.ValidatingLedger.prepareNominatingSet(out agora.consensus.protocol.Data.ConsensusData, ulong, ulong) [0x5570278e3f20]
source/agora/consensus/protocol/Nominator.d:361 @safe bool agora.consensus.protocol.Nominator.Nominator.prepareNominatingSet(out agora.consensus.protocol.Data.ConsensusData) [0x5570278e3d90]
source/agora/consensus/protocol/Nominator.d:427 @safe void agora.consensus.protocol.Nominator.Nominator.checkNominate() [0x5570278e3af0]
submodules/vibe-core/source/vibe/core/core.d:1212 nothrow @safe void vibe.core.core.setTimer(core.time.Duration, void delegate(), bool).__lambda4() [0x557028072980]
submodules/vibe-core/source/vibe/core/core.d:1270 nothrow @safe void vibe.core.core.createTimer(void delegate() nothrow @safe).C.opCall(vibe.core.core.Timer).__lambda2(vibe.core.core.Timer) [0x557028072b00]
submodules/vibe-core/source/vibe/core/task.d:708 nothrow void vibe.core.task.TaskFuncInfo.set!(void delegate(vibe.core.core.Timer) nothrow @safe, vibe.core.core.Timer).set(ref void delegate(vibe.core.core.Timer) nothrow @safe, ref vibe.core.core.Timer).callDelegate(ref vibe.core.task.TaskFuncInfo) [0x5570280b5470]
submodules/vibe-core/source/vibe/core/task.d:737 void vibe.core.task.TaskFuncInfo.call() [0x5570280b2db0]
submodules/vibe-core/source/vibe/core/task.d:403 nothrow void vibe.core.task.TaskFiber.run() [0x5570280b22b0]
??:? fiber_entryPoint [0x7fc34be530a0]
```

**Comment by hewison-chris** [comment]
Looks like there is still an issue as block 432 is second payout block. I will test again locally with latest from v0.x.x

**Comment by hewison-chris** [comment]
It still works for me locally. I think you need release version v0.22 not v0.21.

**Comment by Geod24** [comment]
Fixed, although we still have #2503
